sqlx-sqlite/src/connection/mod.rs

```diff
@@ -544,3 +549,219 @@ impl Statements {
         self.temp = None;
     }
 }
+
+#[cfg(feature = "preupdate-hook")]
+mod preupdate_hook {
+    use super::ConnectionState;
+    use super::LockedSqliteHandle;
+    use super::SqliteOperation;
+    use crate::type_info::DataType;
+    use crate::{SqliteError, SqliteTypeInfo, SqliteValue};
+    use libsqlite3_sys::{
+        sqlite3, sqlite3_preupdate_count, sqlite3_preupdate_depth, sqlite3_preupdate_hook,
+        sqlite3_preupdate_new, sqlite3_preupdate_old, sqlite3_value, sqlite3_value_type, SQLITE_OK,
+    };
+    use sqlx_core::error::Error;
+    use std::ffi::CStr;
+    use std::fmt::Debug;
+    use std::os::raw::{c_char, c_int, c_void};
+    use std::panic::catch_unwind;
+    use std::ptr;
+    use std::ptr::NonNull;
+
+    pub struct PreupdateHookResult<'a> {
+        pub operation: SqliteOperation,
+        pub database: &'a str,
+        pub table: &'a str,
+        pub case: PreupdateCase,
+    }
+
+    pub(crate) struct PreupdateHookHandler(
+        NonNull<dyn FnMut(PreupdateHookResult) + Send + 'static>,
+    );
+    unsafe impl Send for PreupdateHookHandler {}
+
+    /// The possible cases for when a PreUpdate Hook gets triggered. Allows access to the relevant
+    /// functions for each case through the contained values.
+    pub enum PreupdateCase {
+        /// Pre-update hook was triggered by an insert.
+        Insert(PreupdateNewValueAccessor),
+        /// Pre-update hook was triggered by a delete.
+        Delete(PreupdateOldValueAccessor),
+        /// Pre-update hook was triggered by an update.
+        Update {
+            old_value_accessor: PreupdateOldValueAccessor,
+            new_value_accessor: PreupdateNewValueAccessor,
+        },
+        /// This variant is not normally produced by SQLite. You may encounter it
+        /// if you're using a different version than what's supported by this library.
+        Unknown,
+    }
```

Comment:
I'm dubious about the utility of this enum and separating `PreupdateOldValueAccessor` and `PreupdateNewValueAccessor`. If a user wants to handle updates generically, they'd have to match on this and then synthesize some sort of tuple of `(Option<PreupdateOldValueAccessor>, Option<PreupdateNewValueAccessor>)`.

Instead, I'd just merge all this functionality into `PreUpdateHookResult`.

Reply:
Note that it's also undefined behavior to use `sqlite3_preupdate_old_value()` when it's an `INSERT` operation or `_new_value()` when it's a `DELETE` operation, so that also needs to be checked.

Reply:
Done :+1: